clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -1836,7 +1836,19 @@ static TemplateParameterList *GetTemplateParameterList(TemplateDecl *TD) {
   // Make sure we get the template parameter list from the most
   // recent declaration, since that is the only one that is guaranteed to
   // have all the default template argument information.
-  return cast<TemplateDecl>(TD->getMostRecentDecl())->getTemplateParameters();
+  Decl *ND = TD->getMostRecentDecl();
+  // Skip past friend Decls because they are not supposed to contain default
+  // template arguments. Moreover, these declarations may introduce template
+  // parameters living in different template depths than the corresponding
+  // template parameters in TD, causing unmatched constraint substitution.
+  //
+  // C++23 N4950 [temp.param]p12
```

Comment:
nit: this is not a new thing in C++23, I also find it in the C++20 spec N4860. I'd remove the `C++23 N4950`.

And I think this is also a FIXME, clang should diagnose this error case, maybe add the `FIXME` in the comment.

I think this function is not only used for class templates, but also for function templates. For function templates, there is an exception per the standard (temp.param 12): `If a friend function template declaration specifies a default template-argument, that declaration shall be a definition and shall be the only declaration of the function template in the translation unit.`

Reply:
> I think this function is not only used for class templates, but also for function templates. For function templates, there is an exception per the standard (temp.param 12): If a friend function template declaration specifies a default template-argument, that declaration shall be a definition and shall be the only declaration of the function template in the translation unit.

In that case it’d make sense to add a test for a friend function template as well if there isn’t already one to make sure this change doesn’t break that.

Reply:
> nit: this is not a new thing in C++23, I also find it in the C++20 spec N4860. I'd remove the C++23 N4950.

I referenced that number because I'm concerned the paragraphs/wordings could vary across standards. :)
(Maybe I should use an earlier standard file number where this line first appears...?)

Reply:
> I referenced that number because I'm concerned the paragraphs/wordings could vary across standards. :) (Maybe I should use an earlier standard file number where this line first appears...?)

The [stable.names] should never change afaik; as for the paragraphs, it happens to be the same one even in the C++14 standard for this one, so I think it’s fine. You can probably also just reference the C++14 standard if that’s what you prefer.

Reply:
s/C++14/C++20

My bad, I thought this was a different pr for a second

Reply:
> I referenced that number because I'm concerned the paragraphs/wordings could vary across standards. :) (Maybe I should use an earlier standard file number where this line first appears...?)

So apparently, the paragraphs *do* change sometimes, so if you want to cite a specific paragraph, then yeah, I’d include the standard version, but maybe cite the C++20 then in this case if it was added in C++20.

Reply:
> In that case it’d make sense to add a test for a friend function template as well if there isn’t already one to make sure this change doesn’t break that.

I think the change is probably fine for friend function templates with default template parameters: they're unique across one TU, and we have ensured the next declaration we visit is non-null. So in such cases, I think ND would be the equivalent to "the most recent Decl".

> but maybe cite the C++20 then in this case if it was added in C++20.

Finally, I decided to use N3337 (aka C++11) since I couldn't find any related words prior to that draft, although gcc and clang don't allow such default arguments even in 98/03 mode.

Reply:
> I think the change is probably fine for friend function templates with default template parameters: they're unique across one TU, and we have ensured the next declaration we visit is non-null. So in such cases, I think ND would be equivalent to "the most recent Decl".

Judging by the FIXME that’s something we’ll have to come back to anyway, so that’s probably fine for this pr. 

> Finally, I decided to use N3337 (aka C++11) since I couldn't find any related words prior to that draft, although gcc and clang don't allow such default arguments even in 98/03 mode.

Er, I’d just write ‘C++11’ then, because I at least honestly wouldn’t know what N3337 was supposed to be off the top of my head if I just saw that somewhere.

Reply:
> Er, I’d just write ‘C++11’ then, because I at least honestly wouldn’t know what N3337 was supposed to be off the top of my head if I just saw that somewhere.

Oops. I promise I wrote down 'C++11' some time ago, but it seems to have been lost during the copy-paste-format process...

Reply:
> Isn’t ‘N3337’ just the last draft of C++11? Just ‘C++11’ should be enough then?

https://github.com/timsong-cpp/cppwp says N3337 is the one with editorial fixes. So, I think it is.

Reply:
I’d usually not include e.g. ‘N3337’ then unless it’s something that was changed specifically in that version of the standard as opposed to a previous version. If it’s the most up-to-date version, then just e.g. ‘C++11’ in this case would have been enough imo.

Reply:
> At least I can’t find a mention of ‘N3337’ anywhere else in the codebase

Interestingly, I found people tend not to spell out the standard file number throughout the codebase - seemingly, we usually speak `c++(version) [paragraph]/X.Y`.

(I will respect such a convention next time; thanks for reminding me!)

Reply:
> I found people tend not to spell out the standard file number throughout the codebase

Probably beacuse most people don’t have all of them memorised (at least I don’t), whereas e.g. ‘C++11’ makes it pretty clear what we’re actually talking about.